smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java

```diff
@@ -0,0 +1,213 @@
+    private DomainBareJid uploadService;
+    private Long maxFileSize;
+
+    /**
+     * Obtain the HttpFileUploadManager responsible for a connection.
+     *
+     * @param connection the connection object.
+     * @return a HttpFileUploadManager instance
+     */
+    public static synchronized HttpFileUploadManager getInstanceFor(XMPPConnection connection) {
+        HttpFileUploadManager httpFileUploadManager = INSTANCES.get(connection);
+
+        if (httpFileUploadManager == null) {
+            httpFileUploadManager = new HttpFileUploadManager(connection);
+            INSTANCES.put(connection, httpFileUploadManager);
+        }
+
+        return httpFileUploadManager;
+    }
+
+    private HttpFileUploadManager(XMPPConnection connection) {
+        super(connection);
+
+        connection.addConnectionListener(new AbstractConnectionListener() {
+            @Override
+            public void authenticated(XMPPConnection connection, boolean resumed) {
+                // No need to reset the cache if the connection got resumed.
+                if (resumed) {
+                    return;
+                }
+                uploadService = null;
+                maxFileSize = null;
+            }
+        });
+    }
+
+    /**
+     * Discover upload service.
+     *
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @return upload service JID or null if not available
+
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NotConnectedException
+     * @throws InterruptedException
+     * @throws SmackException.NoResponseException
+     */
+    public DomainBareJid discoverUploadService() throws XMPPException.XMPPErrorException, SmackException.NotConnectedException,
+            InterruptedException, SmackException.NoResponseException {
+
+        List<DiscoverInfo> servicesDiscoverInfo = ServiceDiscoveryManager.getInstanceFor(connection())
+                .findServicesDiscoverInfo(SlotRequest.NAMESPACE, true, false);
+
+        if (servicesDiscoverInfo.isEmpty()) {
+            return null;
+        }
+
+        DiscoverInfo discoverInfo = servicesDiscoverInfo.get(0);
+
+        uploadService = discoverInfo.getFrom().asDomainBareJid();
+
+        DataForm dataForm = DataForm.from(discoverInfo);
+        if (dataForm == null) {
+            return uploadService;
+        }
+
+        FormField field = dataForm.getField("max-file-size");
+        if (field == null) {
+            return uploadService;
+        }
+
+        List<String> values = field.getValues();
+        if (!values.isEmpty()) {
+            // TODO: catch NumberFormatException?
+            maxFileSize = Long.valueOf(values.get(0));
+        }
+
+        return uploadService;
+    }
+
+    /**
+     * Get upload service if available.
+     *
+     * Note that you must call {@link HttpFileUploadManager#discoverUploadService} first.
```

Comment:
If I add service discovery in authenticated callback, I have to catch and log exceptions. 
Upload service support can be unknown after error in discovering. 
So, flag should indicate success of discovery operation?